### PR TITLE
add join activity quest/action/criteria

### PR DESCRIPTION
**Pull request by wow890209**
## Why need this change? / Root cause: 
- add join activity quest
- add join activity action
- add join activity criteria
## Changes made:
-
## Test Scope / Change impact:
-

### Summary by OpenAI

**New Features:**
- Introduced `EventJoiningListener` to handle events related to joining and leaving activities.
- Added `Activity` and `Audience` classes for managing activity-related data.
- Implemented `JoinActivityAction` and `JoinActivityCriteria` for functionality related to joining activities.
- Created a new quest, "JoinActivityQuest", with its own criteria and rewards.
- Enhanced user interaction with the addition of `User.presenter` extension function.

> 🎉 With listeners keen and actions bright, 🌟
> 
> We've added quests to your delight. 🚀
> 
> Join an activity, claim your prize, 🏆
> 
> In Utopia, every player flies! 🕊️

### CODE REVIEW

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    override fun onScheduledEventDelete(event: ScheduledEventDeleteEvent) {
+        with(event.scheduledEvent) {
+            val activity = activities.find { it.id == id } ?: return
+            activity.end()
+            log.info("""[activity delete] "activityId" = "$id", "activityName" = "$name"} """)
+        }
+    }
+}
+
+class Activity(
```
Comment (Wally5077) [inline utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt:70]: Q：這是什麼類別，Domain Model？

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/UtopiaGamificationQuestListener.kt [hunk 2]
```diff
@@ -92,9 +94,13 @@ class UtopiaGamificationQuestListener(
...
                 RewardButton.NAME -> handleRewardButtonInteraction(action, args[0])
                 else -> {
-                    playerFulfillMissionsService.execute(action) {
-                        user.claimMissionReward(it)
+
+                    val presenter = object : PlayerFulfillMissionsService.Presenter {
+                        override fun presentClaimMissionReward(mission: Mission) {
+                            user.claimMissionReward(mission)
+                        }
                     }
```
Comment (Wally5077) [inline utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/UtopiaGamificationQuestListener.kt:102]: Ｑ：這邊怎不像前幾個 Service 一樣，擴充成 user 的屬性？

如果要改，就要維持程式碼的統一風格

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/domain/quests/TutorialQuests.kt [hunk 3]
```diff
@@ -193,6 +193,23 @@ val Quests.SendMessageInVoiceChannelQuest: Quest
...
+            """
+            參與任一個活動，並停留40分鐘
+            """.trimIndent()
+
+        reward = Reward(
+            "已完成！",
+            100u,
+        )
+
+        criteria = JoinActivityCriteria("test", 1, 10)
```
Comment (Wally5077) [inline utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/domain/quests/TutorialQuests.kt:212]: Ｑ：JoinActivityCriteria 建構子第一個參數 `test` 是什麼？

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+private fun now() = LocalDateTime.now()
+private fun String.toDate(): LocalDateTime = LocalDateTime.parse(this)
+
+@Component
+class EventJoiningListener(
+    private val playerFulfillMissionsService: PlayerFulfillMissionsService
+) : UtopiaListener() {
+
+    private val activities = mutableListOf<Activity>()
```
Comment (Wally5077) [inline utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt:26]: Q：這該進入 DB 持久化嗎？

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    override fun onGuildVoiceUpdate(event: GuildVoiceUpdateEvent) {
```
Comment (Wally5077) [inline utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt:28]: Q：使用者做什麼事情，會觸發這事件？

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/domain/quests/TutorialQuests.kt (hunk 3 above)
Comment (wow890209) [reply utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/domain/quests/TutorialQuests.kt:212]: 活動的名稱

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt (hunk 5 above)
Comment (wow890209) [reply utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt:28]: 使用者進出語音頻道。

File: utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt (hunk 4 above)
Comment (wow890209) [reply utopia-gamification-quest/src/main/kotlin/tw/waterballsa/utopia/utopiagamificationquest/JoinActivityListener.kt:26]: 不知道

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Pull request merged**